### ISSUE TITLE
Bitcoin integration research

### ISSUE THREAD

**Issue by pfraze** [ssbc/phoenix#402]
@dominictarr and I have been tossing around the notion of supporting BTC transactions in Patchwork. This would allow users to send each other tips.

I've been researching what that would involve. Here are my notes:

 - **Bitcoind** There are pure JS implementations of the crypto, but we probably want to use `bitcoind` from the Bitcoin Core application (available on windows, osx, and linux). It provides a local JSON-RPC interface. It should allow us to integrate with different software and hardware wallets, and it seems likely to be the most hardened codebase. Still researching this.
 - **Key Reuse** Bitcoin docs recommend against reusing a wallet key after a transaction, because it's trivial to track the address' activity. My thoughts on this:
   - Avoiding key reuse means sending a new address to somebody when they want to make a payment. That's a feasible flow for us, but not our bread-and-butter. It's obviously much nicer if we can publish a wallet address for lots of people to tip to.
   - Even with fresh wallet keys, you can still trace transactions. Two ways to counter this: 1) [coinjoins](https://bitcoin.org/en/developer-guide#coinjoin) 2) cash out of BTC into a fiat currency.
   - I suggest, for tips, we don't worry about this, inform users that we reuse the wallet key, and encourage folks to cash out regularly.
   - On the plus side, we'll be able to render how much BTC a tip wallet has received, which is a cool social metric.
 - **Fiat Conversion Rates** There are services which maintain conversion rates to other currencies. It's obviously a variable thing, so, if we want to render in USD, we'll need to ping those services regularly.
 - **SPV vs Full** A user can run their client in a thin mode, called SPV, which doesn't download the entire blockchain (just headers). This mode isn't capable of validating transactions on its own; it requires the user to trust other nodes. You can verify with multiple peers to help with this. Given that the blockchain is 20+ GB, I suggest we default to SPV, and give the user the choice to go to full mode. Eventually, we should be able to overlay SSB's trust graph onto the BTC peer selection.
 - **Hosted Providers** There are services such as Coinbase which will manage the wallet entirely for the user. This is basically like using Paypal. Given the state of device security, putting trust in a hosted solution may appeal to some people, particularly if the wallet-service assumes liability for hacking events. Integration with these providers comes down to supporting their APIs. For reference, Coinbase charges 1% to convert BTC to/from fiats.

**Comment by pfraze** [comment]
Just to clarify about the technology, BTC would be used in Patchwork as a separate protocol. SSB and BTC will not be directly integrated. At most, SSB may treat BTC addresses as first-class references.

**Comment by pfraze** [comment]
Another thought: until we're ready to integrate, we can offer the "copy-and-paste" UX, where we put wallet addresses in the UI but let users copy them into separate wallet software. This would be a good idea while we familiarize ourselves with the BTC landscape.

**Comment by pfraze** [comment]
It turns out that Bitcoin Core doesnt support SPV mode. The canonical implementation for SPV appears to be https://bitcoinj.github.io/. A quick google found https://coinspark.org/sparkbit-wallet/, which supports SPV mode and includes a JSON-RPC interface which should be mostly compatible with `bitcoind`.

**Comment by ralphtheninja** [comment]
@pfraze Newb question: What's Patchwork?

**Comment by pfraze** [comment]
@ralphtheninja new working name for phoenix

**Comment by pfraze** [comment]
Here's a quick shot at what we could do, with a copy/paste UI:

**1)**
![screen shot 2015-07-23 at 3 05 21 pm](https://cloud.githubusercontent.com/assets/1270099/8860970/f7870bf2-314c-11e5-9a6b-e11054fd4075.png)
**2)**
![screen shot 2015-07-23 at 3 05 36 pm](https://cloud.githubusercontent.com/assets/1270099/8860972/f7940ec4-314c-11e5-9fe9-cc5abf50cc06.png)
**3)**
![screen shot 2015-07-23 at 3 09 20 pm](https://cloud.githubusercontent.com/assets/1270099/8860971/f78e4174-314c-11e5-9f23-a020d55375ac.png)

Pretty simple. 1) set a wallet address, 2) click somebody's tip btn, 3) get their wallet address

**Comment by ralphtheninja** [comment]
Very cool! I like the idea of bringing in bitcoin into this a lot. There could be a lot of unforeseen use cases coming out of it as well, e.g. working together in teams and producing open source together and simplifying the process of paying eachother.

**Comment by dominictarr** [comment]
@pfraze there are lots of nuances with bitcoin, and in general the most interesting thing is that there are so many people trying to build blockchain things. so really we need bitcoin as apps i think?

**Comment by pfraze** [comment]
@dominictarr you mean, that we need to provide a bitcoin API to apps?

**Comment by dominictarr** [comment]
@pfraze no i think the apps should provide the bitcoin api. building bitcoin IN will increase maintance surface _massively_

**Comment by pfraze** [comment]
@dominictarr doesn't relying on bitcoind achieve the same thing? Though, Im suggesting for now, we just let users list btc wallet addresses, which is even simpler.

**Comment by dominictarr** [comment]
@pfraze oh sure, so if it's possible to build an app that can connect to bitcoind that would work (easy: it's just local json-rpc over http) but bitcoind is too heavy. I'm guessing there is an svp wallet that provides the same interface though.

**Comment by pfraze** [comment]
@dominictarr yeah, that's the https://coinspark.org/sparkbit-wallet/ I found, though I tried it and found (Java) interface pretty underwhelming
